sdk/tests/conformance2/reading/read-pixels-pack-parameters.html

```diff
@@ -0,0 +1,170 @@
+** MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+*/
+
+-->
+
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<link rel="stylesheet" href="../../resources/js-test-style.css"/>
+<script src="../../js/js-test-pre.js"></script>
+<script src="../../js/webgl-test-utils.js"></script>
+</head>
+<body>
+<canvas id="example" width="4" height="4"></canvas>
+<div id="description"></div>
+<div id="console"></div>
+<script>
+function calculatePaddingBytes(bytesPerPixel, packAlignment, width)
+{
+    var padding = 0;
+    switch (packAlignment) {
+    case 1:
+    case 2:
+    case 4:
+    case 8:
+        padding = (bytesPerPixel * width) % packAlignment;
+        if (padding > 0)
+            padding = packAlignment - padding;
+        return padding;
+    default:
+        testFailed("should not reach here");
+        return;
+    }
+}
+
+function runTestIteration(width, height, packParams)
+{
+    if (!("alignment" in packParams))
+        packParams.alignment = 4;
+    if (!("rowLength" in packParams))
+        packParams.rowLength = 0;
+    if (!("skipPixels" in packParams))
+        packParams.skipPixels = 0;
+    if (!("skipRows" in packParams))
+        packParams.skipRows = 0;
+    debug("Testing width = " + width + ", height = " + height +
+          ", PACK_ALIGNMENT = " + packParams.alignment + ", PACK_ROW_LENGTH = " + packParams.rowLength +
+          ", PACK_SKIP_PIXELS = " + packParams.skipPixels + " , PACK_SKIP_ROWS = " + packParams.skipRows);
+    gl.pixelStorei(gl.PACK_ALIGNMENT, packParams.alignment);
+    gl.pixelStorei(gl.PACK_ROW_LENGTH, packParams.rowLength);
+    gl.pixelStorei(gl.PACK_SKIP_PIXELS, packParams.skipPixels);
+    gl.pixelStorei(gl.PACK_SKIP_ROWS, packParams.skipRows);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.clearColor(expectedColor[0] / 255.0, expectedColor[1] / 255.0, expectedColor[2] / 255.0, expectedColor[3] / 255.0);
+    gl.clear(gl.COLOR_BUFFER_BIT);
+
+    var actualWidth = packParams.rowLength > 0 ? packParams.rowLength : width;
+
+    var bytesPerPixel = 4; // see readPixels' parameters below, the format is gl.RGBA, type is gl.UNSIGNED_BYTE
+    var padding = calculatePaddingBytes(bytesPerPixel, packParams.alignment, actualWidth);
+
+    var size = bytesPerPixel * actualWidth * height + padding * (height - 1);
+    if (size < 0)
+        size = 0;
+    var skipSize = (packParams.skipPixels * bytesPerPixel +
+                    packParams.skipRows * (bytesPerPixel * actualWidth + padding));
+    var buffer = new Uint8Array(size + skipSize);
+    gl.readPixels(0, 0, width, height, gl.RGBA, gl.UNSIGNED_BYTE, buffer);
+    if (packParams.rowLength > 0 && packParams.rowLength < width) {
+        // Revisit this if spec goes the other way.
+        wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION);
+        return;
+    }
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    if (!size)
+        return;
+
+    // Check the last pixel of the last row.
```

Comment:
We may want to consider checking more pixels in these pack alignment tests. It'd be useful to check the first and last pixels of multiple rows.